ts/src/profiler.ts

```diff
@@ -13,47 +13,339 @@
+      packageJson: pjson,
+    };
+    super({parent: new common.Service(serviceConfig, config), baseUrl: '/'});
 
-    this.service = new common.Service(
-        {
-          baseUrl: 'https://cloudprofiler.googleapis.com/v2',
-          scopes: ['https://www.googleapis.com/auth/monitoring.write'],
-        },
-        this.config);
+    this.config = config;
 
     this.logger = new common.logger({
       level: common.logger.LEVELS[config.logLevel as number],
       tag: pjson.name
     });
+
+    this.deployment = {
+      projectId: this.config.projectId,
+      target: this.config.serviceContext.service,
+      labels: {
+        zone: this.config.zone,
+        version: this.config.serviceContext.version
+      }
+    };
+
+    this.profileLabels = {instance: this.config.instance};
+
+    this.profileTypes = [];
+    if (!this.config.disableTime) {
+      this.profileTypes.push(ProfileTypes.Wall);
+      this.timeProfiler = new TimeProfiler(this.config.timeIntervalMicros);
+    }
+    if (!this.config.disableHeap) {
+      this.profileTypes.push(ProfileTypes.Heap);
+      this.heapProfiler = new HeapProfiler(
+          this.config.heapIntervalBytes, this.config.heapMaxStackDepth);
+    }
+  }
+
+  /**
+   * Starts an endless loop to poll profiler server for instructions, and
+   * collects and uploads profiles as requested.
+   * If there is a problem when collecting a profile or uploading a profile to
+   * profiler server, this problem will be logged at the error level and
+   * otherwise ignored.
+   * If there is a problem polling profiler server for instructions
+   * on the type of profile to be collected, this problem will be logged at the
+   * error level and getting profile type will be retried.
+   */
+  start() {
+    this.runLoop();
+  }
+
+  /**
+   * Endlessly polls the profiler server for instructions, and collects and
+   * uploads profiles as requested.
+   */
+  async runLoop(): Promise<void> {
+    const delayMillis = await this.collectProfile();
+
+    // Schedule the next profile.
+    setTimeout(this.runLoop.bind(this), delayMillis).unref();
+  }
+
+  /**
+   * Waits for profiler server to tell it to collect a profile, then collects
+   * a profile and uploads it.
+   *
+   * @return - time, in ms, to wait before asking profiler server again about
+   * collecting another profile.
+   *
+   * TODO: implement backoff and retry. When error encountered in
+   * requestProfile() should be retried when response indicates this request
+   * should be retried or with exponential backoff (up to one hour) if the
+   * response does not indicate when to retry this request.
+   */
+  async collectProfile(): Promise<number> {
+    let prof: RequestProfile;
+    try {
+      prof = await this.requestProfile();
+    } catch (err) {
+      this.logger.error(
+          `Error requesting profile type to be collected: ${err}`);
+      return this.config.backoffMillis;
+    }
+    await this.profileAndUpload(prof);
```

Comment:
If `this.profileAndUpload` throws, would that cause the user application to terminate? That is, I don't see a catch block that would catch this error in the call chain.

Reply:
Errors are caught and logged in profileAndUpload().
I'll switch this, though, for consistency.

Reply:
Done.